kratos/utilities/auxilary_global_master_slave_relation.h

```diff
@@ -0,0 +1,213 @@
+//
+//
+
+#if !defined(AUXILARY_GLOBAL_MASTER_SLAVE_RELATION)
+#define AUXILARY_GLOBAL_MASTER_SLAVE_RELATION
+// System includes
+#include <vector>
+#include <unordered_set>
+#include <iostream>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "includes/process_info.h"
+
+namespace Kratos
+{
+
+/**
+ * @class AuxilaryGlobalMasterSlaveRelation
+ * @ingroup KratosCore
+ * @brief This class stores the information regarding the AuxilaryGlobalMasterSlaveRelation equation.
+ *         Naming convenction is defined like this. (each object of this class will store one equation in the given form
+ *
+ *   SlaveEquationId = w_1*MasterEquationId_1 + w_2*MasterEquationId_2 + ..... + w_n*MasterEquationId_n
+ *
+ *   This stores the condensed form of the MasterSlaveConstraint objects into one object. if only one relation for a slave is added as
+ *   MasterSlaveConstraint then there will only be one entry for master for its corresponding AuxilaryGlobalMasterSlaveRelation.
+ *   Currently this class is designed to hold only one equation. There is only one unique object of this class for each slave. 
+ *
+ *   Future plan is to also make it possible to work with matrices (T) and vectors (for slave and master equation ids and constants)
+ *
+ *
+ *  IMPORTANT : This is not seen by the user. This is a helper data structure which is exists only in the builder and solver. 
+ *
+ * @author Aditya Ghantasala
+ */
+class AuxilaryGlobalMasterSlaveRelation : public IndexedObject
+{
+  public:
+    typedef std::size_t IndexType;
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef std::vector<IndexType> EquationIdVectorType;
+    KRATOS_CLASS_POINTER_DEFINITION(AuxilaryGlobalMasterSlaveRelation);
+
+    // empty constructor and methods to add master and slave independently.
+    AuxilaryGlobalMasterSlaveRelation() : IndexedObject(0)
+    {
+        mConstant = 0.0;
+    }
+
+    AuxilaryGlobalMasterSlaveRelation(IndexType const &rSlaveEquationId) : IndexedObject(rSlaveEquationId)
+    {
+        mConstant = 0.0;
+    }
+
+    /**
+     * Function to get the slave equation Id corresponding to this constraint.
+     * @param Constant the value of the constant to be assigned.
+     */
+    IndexType SlaveEquationId() const { return this->Id(); }
+
+    /**
+     * Function to set the lefthand side of the constraint (the slave dof value)
+     * @param LhsValue the value of the lhs (the slave dof value)
+     */
+    void SetLHSValue(double const &LhsValue)
+    {
+        mLhsValue = LhsValue;
+    }
+
+    /**
+     * Function to update the righthand side of the constraint (the combination of all the master dof values and constants)
+     * @param RhsValue the value of the lhs (the slave dof value)
+     */
+    void SetRHSValue(double const &RhsValue) { mRhsValue = RhsValue; }
+    void UpdateRHSValue(double const &RhsValueUpdate)
```

Comment:
const double, the reference is not needed from what I see

Reply:
variable is copied without a reference isn't  ?

Reply:
Yes, but copying is more efficient in the case of double or integer parameter. A reference creates an extra instruction.

Reply:
oh I see